### PR TITLE
Add front proxy

### PR DESCRIPTION
**Pull request by kylape**
This adds the front proxy into the topology of the deployment manifest.

Instead of terminating TLS at the ingress router, it will instead terminate at the front proxy, and the proxy will route requests to either KCP or the virtual workspaces process.

This also leverages cert-manager to add the ability to authenticate using a certificate.  Groups will be assigned using the organizations listed on the cert.

### CODE REVIEW

**Comment by aojea** [comment]
does this allow to use http2 to the backend?

**Comment by csams** [comment]
@aojea The backends are setup here: https://github.com/kcp-dev/kcp/blob/main/pkg/proxy/proxy.go#L61-L68. We may need to update the `Transport` to enable http2. Would that just be setting `ForceAttemptHTTP2` to true?

File: manifest/kcp-front-proxy.yaml [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        - containerPort: 8443
+        command:
+        - /kcp-front-proxy
+        args:
+        - --secure-port=8443
+        - --tls-private-key-file=/etc/kcp-front-proxy/tls/tls.key
+        - --tls-cert-file=/etc/kcp-front-proxy/tls/tls.crt
+        - --client-ca-file=/etc/kcp-front-proxy/client/tls/ca.crt
+        - --mapping-file=/etc/kcp-front-proxy/config/path-mapping.yaml
+        - --v=4
```
Comment (csams) [inline manifest/kcp-front-proxy.yaml:183]: Let's change this to 6 so the proxy will log basic info about requests to the backends.

**Review by csams — CHANGES_REQUESTED** [review]

**Comment by kylape** [comment]
We can try out the changes on https://github.com/kcp-dev/kcp/pull/873 with this manifest to see if http2 works.

**Comment by aojea** [comment]
> @aojea The backends are setup here: https://github.com/kcp-dev/kcp/blob/main/pkg/proxy/proxy.go#L61-L68. We may need to update the `Transport` to enable http2. Would that just be setting `ForceAttemptHTTP2` to true?

I would build the transport differently, the ReverseProxy uses http.DefaultTransport by default https://pkg.go.dev/net/http/httputil#ReverseProxy
```
// DefaultTransport is the default implementation of Transport and is
// used by DefaultClient. It establishes network connections as needed
// and caches them for reuse by subsequent calls. It uses HTTP proxies
// as directed by the $HTTP_PROXY and $NO_PROXY (or $http_proxy and
// $no_proxy) environment variables.
var DefaultTransport RoundTripper = &Transport{
	Proxy: ProxyFromEnvironment,
	DialContext: defaultTransportDialContext(&net.Dialer{
		Timeout:   30 * time.Second,
		KeepAlive: 30 * time.Second,
	}),
	ForceAttemptHTTP2:     true,
	MaxIdleConns:          100,
	IdleConnTimeout:       90 * time.Second,
	TLSHandshakeTimeout:   10 * time.Second,
	ExpectContinueTimeout: 1 * time.Second,
}
```

initializing it https://github.com/kcp-dev/kcp/blob/main/pkg/proxy/proxy.go#L61-L68 this way you are overriding some sane defaults. I would initialize it as:
```go
        transport = http.DefaultTransport
        transport.TLSClientConfig = &tls.Config{
			Certificates: []tls.Certificate{cert},
			RootCAs:      caCertPool,
		},
	proxy := httputil.NewSingleHostReverseProxy(target)
	proxy.Transport = transport
```

**Comment by kylape** [comment]
Note that this is checking in an Openshift-specific resource (Route).  This is a short term solution; we intend to have a more generic manifest available for the community that doesn't assume any particular distribution.

**Pull request merged**